### PR TITLE
Undo deletion iOS UI

### LINKED ISSUE #2540: Undo time entry deletion UI
- Use the same animation to show and hide the snackbar: #1538

### Undo when the timer is not running
toggl - ios - timer - undo 1x

Zeplin: https://zpl.io/b6MDQrR

### Undo when the timer is running
toggl - ios - timer - undo running entry 1x

Zeplin: https://zpl.io/agvk9AZ

### Tappable area: 8pt padding from left and right; 10pt from to and bottom
screen shot 2018-07-02 at 13 07 48

### PR DESCRIPTION
**Pull request by Odrakir**
## What's this?
This adds the Snack bar for UNDO deletion. It's reusable enough so it can also be used in the feedback ui and password reset (or wherever else needed) with minimal or non modification. Ping @khoi & @arturs105 if interested.

It decides where to put the buttons dynamically depending on the text length and view size. So you can have both the following views with the same class.

captura de pantalla 2018-08-02 a las 11 40 51
captura de pantalla 2018-08-02 a las 11 42 02

Also, it has a public property to set the bottom anchor, which can be changed in run time and it adapts to the new anchor.

### Relationships
Closes #2540 #2750 

## Why do we want this?
The next step is the whole undo logic, but this PR brings the UI functionality needed for that.

## How is it done?
It's just a UIView subclass with some extra methods. The default colors and fonts are taken from the xib file, but they can be changed programmatically.

## Review hints
In order to test it you can replace the Start button behavior so it shows this bar.

In line 148 of `MainViewController.cs` do:
```c#
//bindingSet.Bind(StartTimeEntryButton).To(vm => vm.StartTimeEntryCommand);
var snackBar = SnackBar.Create();
snackBar.SnackBottomAnchor = StartTimeEntryButton.TopAnchor;
snackBar.AddButton("UNDO", () =>
{
    Console.WriteLine("Tap");
});
snackBar.SetTimer(3, () =>
{
    Console.WriteLine("Fin");
});
StartTimeEntryButton.Tapped()
    .VoidSubscribe(() => { snackBar.Show("Time entry was deleted", View); });

ViewModel.IsTimeEntryRunning
    .Subscribe(running =>
    {
        snackBar.SnackBottomAnchor = running ? CurrentTimeEntryCard.TopAnchor : StartTimeEntryButton.TopAnchor;
    })
    .DisposedBy(disposeBag);
```

## :squid: Permissions
Go ahead.

### CODE REVIEW

File: Toggl.Daneel/Views/SnackBar.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                if (!showing) return;
+                Hide();
+                onTimer();
+            };
+        }
+
+        public void Show(string text, UIView superView)
+        {
+            if (Superview != null)
+            { //Only show it once
```
Comment (simonrozsival) [inline Toggl.Daneel/Views/SnackBar.cs]: 👀

File: Toggl.Daneel/Views/SnackBar.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        }
+
+        public void Hide()
+        {
+            showing = false;
+            UIView.Animate(
+                0.3,
+                () =>
+                {
+                    Transform = CGAffineTransform.MakeTranslation(0, 100);
```
Comment (simonrozsival) [inline Toggl.Daneel/Views/SnackBar.cs]: Magic 💯 - you use this constant twice for translation, please give it a name.

File: Toggl.Daneel/Views/SnackBar.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                    buttonPosition = ButtonPositionType.Bottom;
+                    configureLayout();
+                    LayoutIfNeeded(); // Seems silly, but this one is needed too
+                }
+            }
+        }
+
+        private void configure()
+        {
+            Layer.CornerRadius = 10;
```
Comment (simonrozsival) [inline Toggl.Daneel/Views/SnackBar.cs]: A named constant please!

File: Toggl.Daneel/Views/SnackBar.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        public override void LayoutSubviews()
+        {
+            base.LayoutSubviews();
+
+            if (buttonPosition == ButtonPositionType.Right)
+            {
+                LayoutIfNeeded(); // Seems silly, but I need this line so the label returns the correct width
+                var labelWidth = label.Frame.Width;
+                var requiredWidth = ((NSString)label.Text).GetBoundingRect(new CGSize(2000, 2000),
```
Comment (simonrozsival) [inline Toggl.Daneel/Views/SnackBar.cs]: Why 2000x2000?

File: Toggl.Daneel/Views/SnackBar.cs (hunk 1 above)
Comment (Odrakir) [reply Toggl.Daneel/Views/SnackBar.cs]: I'm not a big fan of commenting everything, but I think that comment makes sense.

Comment (simonrozsival) [reply Toggl.Daneel/Views/SnackBar.cs]: I don't mind the comment, the formatting seems off though. I should have been more specific the first time.

**Comment by khoi** [comment]
Probably going to close #2750 as well

**Comment by simonrozsival** [comment]
How is this gonna work when you delete several TEs while the snack bar is visible - will tapping the UNDO button reverse the deletion of the first or the last TE?

**Comment by Odrakir** [comment]
According to the issue deleting a new entry will delete the previous one for real. Still not sure how I'm going to handle that, but probably not in the UI, so the old snack bar will just dismiss some how.

**Pull request merged**